src/nupic/research/connections.py

```diff
@@ -342,47 +372,40 @@ def createSynapse(self, segment, presynapticCell, permanence):
     while self.numSynapses(segment) >= self.maxSynapsesPerSegment:
       self.destroySynapse(self._minPermanenceSynapse(segment))
 
-    if segment._numDestroyedSynapses > 0:
-      synapse = next(s for s in segment._synapses if s._destroyed)
+    idx = len(segment._synapses)
+    synapse = Synapse(segment, presynapticCell, permanence,
+                      self._nextSynapseOrdinal)
+    self._nextSynapseOrdinal += 1
+    segment._synapses.add(synapse)
 
-      synapse._destroyed = False
-      segment._numDestroyedSynapses -= 1
+    self._synapsesForPresynapticCell[presynapticCell].add(synapse)
 
-      synapse.presynapticCell = presynapticCell
-      synapse.permanence = permanence
+    self._numSynapses +=1
 
-    else:
-      idx = len(segment._synapses)
-      synapse = Synapse(segment, idx, presynapticCell, permanence)
-      segment._synapses.append(synapse)
+    return synapse
 
-    self._synapsesForPresynapticCell[presynapticCell].append(synapse)
-    self._numSynapses += 1
 
-    return synapse
+  def _removeSynapseFromPresynapticMap(self, synapse):
+    presynapticSynapses = \
```

Comment:
Yeah, the words "presynaptic" and "synapse" are long and each appear three times in this statement. I'll change the list name to "inputSynapses".